### PR TITLE
Fix non-symmetry bug in FingerprintChecker

### LINKED ISSUE #1437: CrossCheck fingerprint is not symmetric (A vs. B != B vs. A)
## Bug Report
It has been reported to me (and I've confirmed) that CrosscheckFingerprints is not symmetric when comparing A to B and B to A. This seems to be due to two separate issues:

1. There's a bug that was introduced during refactoring whereby the prior isn't applied correctly 
2. The abundance of applications of the log function seems to cause floating point errors which depend on the order of operations and therefore are not symmetric.

While 1. is relatively easy to fix (I have a fix on a branch) the tests are failing due to 2. which I do not plan to fix. I will likely need to introduce some leniency in the test due to 2.

### PR DESCRIPTION
**Pull request by yfarjoun**
Fixes a bug in the fingerprinting code that resulted in non-symmetric results

Results: LOD(a,b) != LOD(b,a)

The problem was that the __normlized__ posterior was used as the prior in the denominator of the LOD calculation.
The correct thing to do is to use the __unnormalized__ posterior array which is what this commit does.

Other changes include:

- more tests
- cleanup

Fixes #1437 

### Description

_Give your PR a **concise** yet **descriptive** title_
_Please explain the changes you made here._
_Explain the **motivation** for making this change. What existing problem does the pull request solve?_
_Mention any issues fixed, addressed or otherwise related to this pull request, including issue numbers or hard links for issues in other repos._
_You can delete these instructions once you have written your PR description._

----

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

#### Content
- [ ] Added or modified tests to cover changes and any new functionality
- [ ] Edited the README / documentation (if applicable)
- [ ] All tests passing on Travis

#### Review
- [ ] Final thumbs-up from reviewer
- [ ] Rebase, squash and reword as applicable

For more detailed guidelines, see https://github.com/broadinstitute/picard/wiki/Guidelines-for-pull-requests

### CODE REVIEW

File: src/main/java/picard/fingerprint/HaplotypeProbabilities.java [hunk 1]
```diff
@@ -213,16 +220,16 @@ public double shiftedLogEvidenceProbabilityGivenOtherEvidence(final HaplotypePro
...
+
+            P(hap|x,y) = P(x|hap,y) P(hap|y) / P(x|y)
+                       = P(x | hap) * P(hap | y) / P(x)
+                         likelihood * other.posterior
+
+                       = P(x|hap) P(y|hap) P(hap)/P(x)P(y)
+                       = A P(x| hap) P(y| hap) P(hap)  # where A is an unknown scaling factor
          */
-        return shiftedLogEvidenceProbabilityUsingGenotypeFrequencies(otherHp.getPosteriorProbabilities());
+        return shiftedLogEvidenceProbabilityUsingGenotypeFrequencies(otherHp.getPosteriorLikelihoods());
```
Comment (yfarjoun) [inline src/main/java/picard/fingerprint/HaplotypeProbabilities.java]: For the reviewer, this is the bug fix....most of everything else is window dressing.

File: src/main/java/picard/util/MathUtil.java [hunk 2]
```diff
@@ -285,6 +324,38 @@ public static int indexOfMin(final double[] nums) {
         return pNormalizeLogProbability(likelihoods);
     }
 
+    /**
+     * Takes a complete set of mutually exclusive logLikelihoods and converts them to probabilities
+     * with no rescaling. Will throw if underflow is detected (if all the likelihoods are less than -300)
+     */
+    public static double[] getProbabilityFromLog(final double[] lLikelihood) {
```
Comment (pshapiro4broad) [inline src/main/java/picard/util/MathUtil.java]: New public APIs should have tests.

File: src/main/java/picard/util/MathUtil.java [hunk 3]
```diff
@@ -441,7 +556,34 @@ public static double klDivergance(double[] measured, double[] distribution) {
...
+     * A small utility function to choose n random elements (un-shuffled) from a list
+     *
+     * @param list A list of elements
+     * @param n    a number of elements requested from list
+     * @return a list of n randomly chosen (but in the original order) elements from list.
+     * If the list has less than n elements it is returned in its entirety.
+     */
+    public static <T> List<T> randomSublist(final List<T> list, final int n) {
+        int availableElements = list.size();
+        if (availableElements <= n) return list;
```
Comment (pshapiro4broad) [inline src/main/java/picard/util/MathUtil.java]: Would be nice to reformat this too

File: src/main/java/picard/util/MathUtil.java [hunk 4]
```diff
@@ -441,7 +556,34 @@ public static double klDivergance(double[] measured, double[] distribution) {
...
+
+        int stillNeeded = n;
+        final Random rg = new Random();
```
Comment (pshapiro4broad) [inline src/main/java/picard/util/MathUtil.java]: Any reason to not use `Math.random()` here?

**Comment by yfarjoun** [comment]
thanks @pshapiro4broad 

responded to your review.

File: src/main/java/picard/fingerprint/FingerprintChecker.java [hunk 5]
```diff
@@ -88,6 +89,8 @@
     public static final int DEFAULT_MINIMUM_MAPPING_QUALITY = 10;
     public static final int DEFAULT_MINIMUM_BASE_QUALITY = 20;
     public static final int DEFAULT_MAXIMAL_PL_DIFFERENCE = 30;
+    // used sometimes to subset loci. Fix the random seed so that the results are deterministic
+    public static final Random random = new Random(42);
```
Comment (pshapiro4broad) [inline src/main/java/picard/fingerprint/FingerprintChecker.java]: Unless there's a need to use it elsewhere, I would make this `private`.

File: src/main/java/picard/fingerprint/FingerprintChecker.java [hunk 6]
```diff
@@ -633,7 +634,7 @@ private FingerprintIdDetails createUnknownFP(final Path samFile, final SAMRecord
                 final Snp snp = this.haplotypes.getSnp(info.getSequenceName(), info.getPosition());
 
                 // randomly select locusMaxReads elements from the list
-                final List<SamLocusIterator.RecordAndOffset> recordAndOffsetList = randomSublist(info.getRecordAndPositions(), locusMaxReads);
+                final List<SamLocusIterator.RecordAndOffset> recordAndOffsetList = MathUtil.randomSublist(info.getRecordAndOffsets(), locusMaxReads,random);
```
Comment (pshapiro4broad) [inline src/main/java/picard/fingerprint/FingerprintChecker.java]: whitespace
```suggestion
                final List<SamLocusIterator.RecordAndOffset> recordAndOffsetList = MathUtil.randomSublist(info.getRecordAndOffsets(), locusMaxReads, random);
```

File: src/test/java/picard/util/MathUtilTest.java [hunk 7]
```diff
@@ -84,6 +88,19 @@ private void assertEquals(final double[] actual, final double[] expected) {
         }
     }
 
+    @Test
+    public void testRandomSublist() {
+        final Random random = new Random();
+        final List<Integer> list = new ArrayList<>();
```
Comment (pshapiro4broad) [inline src/test/java/picard/util/MathUtilTest.java]: Could write this as
```
        final List<Integer> list = Arrays.asList(1, 2, 3);
```

File: src/test/java/picard/fingerprint/FingerprintCheckerTest.java [hunk 8]
```diff
@@ -143,27 +160,30 @@ public void testTerminateOnBadFile() {
...
+                {na12891_r2, na12891_noRg, true, false},
+                {na12892_r2, na12891_noRg, false, false},
         };
     }
 
     @Test(dataProvider = "checkFingerprintsSamDataProvider")
-    public void testCheckFingerprintsSam(final File samFile1, final File samFile2, final boolean expectedMatch, final boolean silent) {
+    public void testCheckFingerprintsSam(final File samFile1, final File samFile2, final boolean expectedMatch, final boolean silent) throws IOException {
+
+        final File metricsFile = File.createTempFile("crosscheck",".crosscheck_metrics");
```
Comment (pshapiro4broad) [inline src/test/java/picard/fingerprint/FingerprintCheckerTest.java]: whitespace
```suggestion
        final File metricsFile = File.createTempFile("crosscheck", ".crosscheck_metrics");
```

File: src/test/java/picard/fingerprint/FingerprintCheckerTest.java [hunk 9]
```diff
@@ -173,9 +193,29 @@ public void testCheckFingerprintsSam(final File samFile1, final File samFile2, f
                 "I=" + samFile2.getAbsolutePath(),
                 "VALIDATION_STRINGENCY=" + (silent ? "SILENT" : "LENIENT"),
                 "CROSSCHECK_BY=FILE",
+                "OUTPUT="+metricsFile.getAbsolutePath()
```
Comment (pshapiro4broad) [inline src/test/java/picard/fingerprint/FingerprintCheckerTest.java]: whitespace
```suggestion
                "OUTPUT=" + metricsFile.getAbsolutePath()
```

File: src/test/java/picard/fingerprint/FingerprintCheckerTest.java [hunk 10]
```diff
@@ -124,13 +124,30 @@ public void testCheckFingerprintsVcf(final File vcfFile, final File genotypesFil
...
 
+
+    @Test(dataProvider = "checkFingerprintsVcfDataProvider")
+    public void testFingerprintSwapEqual(final File vcfFile, final File genotypesFile, final String observedSampleAlias, final String expectedSampleAlias,
+                                   final double llExpectedSample, final double llRandomSample, final double lodExpectedSample) {
+        final FingerprintChecker fpChecker = new FingerprintChecker(SUBSETTED_HAPLOTYPE_DATABASE_FOR_TESTING);
+        final Map<FingerprintIdDetails, Fingerprint> fp1Map = fpChecker.fingerprintVcf(vcfFile.toPath());
+        final Map<FingerprintIdDetails, Fingerprint> fp2Map = fpChecker.fingerprintVcf(genotypesFile.toPath());
+
+        for(Fingerprint fp1:fp1Map.values()){
```
Comment (pshapiro4broad) [inline src/test/java/picard/fingerprint/FingerprintCheckerTest.java]: whitespace
```suggestion
        for (Fingerprint fp1 : fp1Map.values()) {
```

File: src/test/java/picard/fingerprint/FingerprintCheckerTest.java [hunk 11]
```diff
@@ -124,13 +124,30 @@ public void testCheckFingerprintsVcf(final File vcfFile, final File genotypesFil
...
+            for(Fingerprint fp2:fp2Map.values()) {
+                final MatchResults matchResults12 = FingerprintChecker.calculateMatchResults(fp1, fp2);
+                final MatchResults matchResults21 = FingerprintChecker.calculateMatchResults(fp2, fp1);
+                compareDoubleWithAccuracy(matchResults12.getLOD(),matchResults21.getLOD(),1e-10);
```
Comment (pshapiro4broad) [inline src/test/java/picard/fingerprint/FingerprintCheckerTest.java]:
```suggestion
                compareDoubleWithAccuracy(matchResults12.getLOD(),matchResults21.getLOD(), 1e-10);
```

**Comment by yfarjoun** [comment]
WOOT!

**Comment by dinvlad** [comment]
Amazing! Thank you so much @yfarjoun and @pshapiro4broad!

Is a new release underway? :P

**Pull request merged**